src/ripple/core/JobCounter.h

```diff
@@ -0,0 +1,188 @@
+                "JobHandler must be callable with Job&");
+
+    public:
+        CountedJob () = delete;
+
+        CountedJob (CountedJob const& rhs)
+        : jobCounter_ (rhs.jobCounter_)
+        , jobHandler_ (rhs.jobHandler_)
+        {
+            jobCounter_.adjustCount (1);
+        }
+
+        CountedJob (CountedJob&& rhs)
+        : jobCounter_ (rhs.jobCounter_)
+        , jobHandler_ (std::move (rhs.jobHandler_))
+        {
+            jobCounter_.adjustCount (1);
+        }
+
+        CountedJob (JobCounter& jobCounter, JobHandler&& jobHandler)
+        : jobCounter_ (jobCounter)
+        , jobHandler_ (std::move (jobHandler))
+        {
+            jobCounter_.adjustCount (1);
+        }
+
+        CountedJob& operator=(CountedJob const& rhs) = delete;
+        CountedJob& operator=(CountedJob&& rhs) = delete;
+
+        ~CountedJob ()
+        {
+            jobCounter_.adjustCount (-1);
+        }
+
+        void operator ()(Job& job)
+        {
+            return jobHandler_ (job);
+        }
+    };
+
+public:
+    JobCounter() = default;
+    // Not copyable because outstanding counts would be hard to sort out.
+    JobCounter (JobCounter const&) = delete;
+    // Not movable because in-flight counted Jobs have a pointer to this.
+    JobCounter (JobCounter&&) = delete;
+
+    JobCounter& operator=(JobCounter const&) = delete;
+
+    /** Destructor verifies all in-flight jobs are complete. */
+    ~JobCounter()
+    {
+        join();
+    }
+
+    /** Returns once all counted in-flight Jobs are destroyed. */
+    void join()
+    {
+        std::unique_lock<std::mutex> lock {mutex_};
+        waitForJobs_ = true;
+        if (jobCount_ > 0)
+        {
+            allJobsDoneCond_.wait (lock, [this] { return jobCount_ == 0; });
+        }
+    }
+
+    /** Wrap the passed lambda with a reference counter.
+
+        @param jobHandler Lambda that accepts a Job& parameter and returns void.
+        @return If join() has been called returns boost::none.  Otherwise
+                returns a boost::optional that wraps jobHandler with a
+                reference counter.
+    */
+    template <class JobHandler>
+    boost::optional<CountedJob<JobHandler>>
+    wrap (JobHandler&& jobHandler)
+    {
+        // We must construct the wrapped Job *first* and then check
+        // whether join() has been called.  If join() has already been
+        // called then return boost::none.  This behavior may possibly
+        // cause allJobsDoneCond_ to notify() multiple times at shutdown,
+        // but that should cause no harm.
+        //
+        // We cannot first check if join() has been called and then
+        // construct the wrapped job if we were not previously joined.
+        // That would introduce a timing window.
+        boost::optional<CountedJob<JobHandler>> ret {
+            CountedJob<JobHandler> {
+                *this, std::forward<JobHandler> (jobHandler)}};
```

Comment:
Right.  All of the current uses of `wrap()` involve `rvalue` lambdas.  You have an excellent point that `wrap()` is not restrictive enough in the types that it takes.

I like your suggestion of using `remove_reference_t`.  That never occurred to me (obviously).

Also, since we're not overloading `wrap()`, a `static_assert` might do the trick.  Perhaps something like:
```
static_assert (std::is_rvalue_reference<decltype (handler)>::value,
    "JobCounter::wrap() only supports rvalue lambdas.");
```
What's your preference?